Requesting computing resources
```shell
sinteractive -c 28 -t 01:00:00 -J Calling -A PAS2658
```
Get into the right wd
```shell
cd /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Calling
```
Sorting with Picard
```shell
# For Nd-1 short
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/picard.sif picard -Xmx10g SortSam \
I=../Alignments/Nd-1_short.bam \
O=Nd-1_short_sorted.bam \
VALIDATION_STRINGENCY=LENIENT \
SORT_ORDER=coordinate \
MAX_RECORDS_IN_RAM=3000000 \
CREATE_INDEX=True  
# For Nd-1 long
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/picard.sif picard -Xmx100g SortSam \
I=../Alignments/Nd-1_long.bam \
O=Nd-1_long_sorted.bam \
VALIDATION_STRINGENCY=LENIENT \
SORT_ORDER=coordinate \
MAX_RECORDS_IN_RAM=3000000 \
CREATE_INDEX=True  
# For Ler-0 short
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/picard.sif picard -Xmx10g SortSam \
I=../Alignments/Ler-0_short.bam \
O=Ler-0_short_sorted.bam \
VALIDATION_STRINGENCY=LENIENT \
SORT_ORDER=coordinate \
MAX_RECORDS_IN_RAM=3000000 \
CREATE_INDEX=True  
```
Obtaining some stats from the BAM files
```shell
# For Nd-1 short
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools flagstat Nd-1_short_sorted.bam
# For Nd-1 long
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools flagstat Nd-1_long_sorted.bam
# For Ler-0 short
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools flagstat Ler-0_short_sorted.bam
```
Marking duplicates
```shell
    # For Nd-1 short
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/picard.sif picard -Xmx100g MarkDuplicates\
    I=Nd-1_short_sorted.bam \
    O=Nd-1_short_sorted_dup.bam \
    METRICS_FILE=marked_dup_metrics_Nd-1_short.txt
    # For Nd-1 long
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/picard.sif picard -Xmx100g MarkDuplicates\
    I=Nd-1_long_sorted.bam \
    O=Nd-1_long_sorted_dup.bam \
    METRICS_FILE=marked_dup_metrics_Nd-1_long.txt
    # For Ler-0 short
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/picard.sif picard -Xmx100g MarkDuplicates\
    I=Ler-0_short_sorted.bam \
    O=Ler-0_short_sorted_dup.bam \
    METRICS_FILE=marked_dup_metrics_Ler-0_short.txt
```
Adding read group tags (@RG) lines in the header of BAM files
```shell
# for Nd-1 short
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools addreplacerg -@ 28 -r '@RG\tID:samplename\tSM:samplename' Nd-1_short_sorted_dup.bam -o Nd-1_short_sorted_dup_grouped.bam --write-index
# for Nd-1 long
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools addreplacerg -@ 28 -r '@RG\tID:samplename\tSM:samplename' Nd-1_short_sorted_dup.bam -o Nd-1_long_sorted_dup_grouped.bam --write-index
# for Ler-0 short
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools addreplacerg -@ 28 -r '@RG\tID:samplename\tSM:samplename' Ler-0_short_sorted_dup.bam -o Ler-0_short_sorted_dup_grouped.bam --write-index
# for Ler-0 long
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools addreplacerg -@ 28 -r '@RG\tID:samplename\tSM:samplename' Ler-0_long_sorted_dup.bam -o Ler-0_long_sorted_dup_grouped.bam --write-index
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools addreplacerg -@ 28 -r '@RG\tID:samplename\tSM:samplename' Ler-0_long_sorted.bam -o Ler-0_long_sorted_grouped.bam --write-index
```
Creating reference sequence index and dictionary
```shell
# index
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools faidx /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Reference_Genomes/TAIR10.fasta
# dictionary
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4.sif gatk CreateSequenceDictionary -R /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Reference_Genomes/TAIR10.fasta
```

Applying HaplotypeCaller to do variant calling inchromosome 4 only
```shell
tee -a Haplotype_Calling.sh <<EOF
#!/bin/bash
#SBATCH -J ondemand/sys/myjobs/basic_sequential
#SBATCH --job-name HaplotypeCaller
#SBATCH --time=02:00:00
#SBATCH --ntasks=28
#SBATCH --exclusive
#SBATCH --mail-type=BEGIN,END,FAIL
#SBATCH --account=PAS2658
# Applying HaplotypeCaller to do variant calling inchromosome 4 only
cd /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Calling
# Calling Haplotypes for Nd-1 short
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4.sif gatk --java-options "-Xmx120g" HaplotypeCaller \
    -I Nd-1_short_sorted_dup_grouped.bam \
    --native-pair-hmm-threads 28 \
    -ERC GVCF \
    -R /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Reference_Genomes/TAIR10.fasta \
    -L 4 \
    -O Nd-1_short.g.vcf.gz
# For Nd-1 long
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4.sif gatk --java-options "-Xmx120g" HaplotypeCaller \
    -I Nd-1_long_sorted_dup_grouped.bam \
    --native-pair-hmm-threads 28 \
    -ERC GVCF \
    -R /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Reference_Genomes/TAIR10.fasta \
    -L 4 \
    -O Nd-1_long.g.vcf.gz
# For Ler-0 short
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4.sif gatk --java-options "-Xmx120g" HaplotypeCaller \
    -I Ler-0_short_sorted_dup_grouped.bam \
    --native-pair-hmm-threads 28 \
    -ERC GVCF \
    -R /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Reference_Genomes/TAIR10.fasta \
    -L 4 \
    -O Ler-0_short.g.vcf.gz
# For Ler-0 long
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4_Spark.sif gatk --java-options "-Xmx120g" HaplotypeCallerSpark \
    -I LLer-0_long_sorted_grouped_dup.bam \
    --native-pair-hmm-threads 28 \
    -R /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Reference_Genomes/TAIR10.fasta \
    --read-shard-size 67000 \
    -L 4 \
    -ERC GVCF \
    -O Ler-0_2_long.g.vcf.gz
EOF
sbatch Haplotype_Calling.sh
```

Combine VGCFs
```shell
cd /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Calling
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4.sif gatk --java-options "-Xmx120g" CombineGVCFs \
    -R /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Reference_Genomes/TAIR10.fasta \
    -V Nd-1_short.g.vcf.gz \
    -V Nd-1_long.g.vcf.gz \
    -V Ler-0_short.g.vcf.gz \
    -O Nd1_Ler0.g.vcf.gz
```

Apply GenotypeGVCFs
```shell
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4.sif gatk --java-options "-Xmx120g" GenotypeGVCFs \
    -R /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Reference_Genomes/TAIR10.fasta \
    -V Nd1_Ler0.g.vcf.gz \
    -O Nd1_Ler0.vcf.gz
```

Using VCFtools for filtering the VCF
```shell
# Keeping only polymorphic SNPs with a minimum depth of 10, no missing data, and quality of 38
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/vcftools.sif vcftools --gzvcf Nd1_Ler0.vcf.gz --remove-indels --maf 0.1 --minDP 10 --max-missing 1  --minQ 38 --recode --out Nd1_Ler0_cleaned
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/vcftools.sif vcftools --vcf Nd1_Ler0_cleaned.recode.vcf
```

Using VariantsToTable to obtain a tsv file readable in software like Microsoft Excel
```shell
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4.sif gatk VariantsToTable \
    -R /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Reference_Genomes/TAIR10.fasta \
    -V Nd1_Ler0_cleaned.recode.vcf \
    -F CHROM -F POS -F FILTER -F TYPE -GF AD -GF DP \
    --show-filtered \
    -O Nd1_Ler0_variants.tsv
```
