Requesting computing resources
```shell
sinteractive -c 28 -t 01:00:00 -J Sooty -A PAS2658
```
Get into the righ wd
```shell
cd /fs/scratch/PASXXXX/Your_WD/Variant_Calling/
mkdir Sooty_Blotch
cd /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Sooty_Blotch
```
Copying reference and BAM files from RNA-seq
Get into the righ wd
```shell
cp /fs/scratch/PAS2658/Jonathan/Gene_Expression/Reference/Sooty_Blotch/P_fructicola.fasta .
cp /fs/scratch/PAS2658/Jonathan/Gene_Expression/Rsubread/Sooty_Blotch/*bam .
```
Sorting with Picard
```shell
# For PD_05d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/picard.sif picard -Xmx10g SortSam \
I=PD_05d.bam \
O=PD_05d_sorted.bam \
VALIDATION_STRINGENCY=LENIENT \
SORT_ORDER=coordinate \
MAX_RECORDS_IN_RAM=3000000 \
CREATE_INDEX=True  
# For PD_15d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/picard.sif picard -Xmx100g SortSam \
I=PD_15d.bam \
O=PD_15d_sorted.bam \
VALIDATION_STRINGENCY=LENIENT \
SORT_ORDER=coordinate \
MAX_RECORDS_IN_RAM=3000000 \
CREATE_INDEX=True  
# For PEG15d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/picard.sif picard -Xmx10g SortSam \
I=PEG15d.bam \
O=PEG15d_sorted.bam \
VALIDATION_STRINGENCY=LENIENT \
SORT_ORDER=coordinate \
MAX_RECORDS_IN_RAM=3000000 \
CREATE_INDEX=True  
```
Obtaining some stats from the BAM files
```shell
# For PD_05d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools flagstat PD_05d_sorted.bam
# For PD_15d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools flagstat PD_15d_sorted.bam
# For PEG15d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools flagstat PEG15d_sorted.bam
```
Marking duplicates
```shell
    # For PD_05d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/picard.sif picard -Xmx100g MarkDuplicates\
    I=PD_05d_sorted.bam \
    O=PD_05d_sorted_dup.bam\
    METRICS_FILE=marked_dup_metrics_PD_05d.txt
    # For PD_15d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/picard.sif picard -Xmx100g MarkDuplicates\
    I=PD_15d_sorted.bam \
    O=PD_15d_sorted_dup.bam \
    METRICS_FILE=marked_dup_metrics_PD_15d.txt
    # For PEG15d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/picard.sif picard -Xmx100g MarkDuplicates\
    I=PEG15d_sorted.bam \
    O=PEG15d_sorted_dup.bam \
    METRICS_FILE=marked_dup_metrics_PEG15d.txt
```
Adding read group tags (@RG) lines in the header of BAM files
```shell
# For PD_05d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools addreplacerg -@ 28 -r '@RG\tID:samplename\tSM:samplename' PD_05d_sorted_dup.bam -o PD_05d_sorted_dup_grouped.bam --write-index
# For PD_15d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools addreplacerg -@ 28 -r '@RG\tID:samplename\tSM:samplename' PD_15d_sorted_dup.bam -o PD_15d_sorted_dup_grouped.bam --write-index
# For PEG15d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools addreplacerg -@ 28 -r '@RG\tID:samplename\tSM:samplename' PEG15d_sorted_dup.bam -o PEG15d_sorted_dup_grouped.bam --write-index
```
Creating reference sequence index and dictionary
```shell
# index
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/samtools.sif samtools faidx P_fructicola.fasta
# dictionary
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4.sif gatk CreateSequenceDictionary -R P_fructicola.fasta
```

Applying HaplotypeCaller to do variant calling inchromosome 4 only
```shell
tee -a Haplotype_Calling.sh <<EOF
#!/bin/bash
#SBATCH -J ondemand/sys/myjobs/basic_sequential
#SBATCH --job-name HaplotypeCaller
#SBATCH --time=02:00:00
#SBATCH --ntasks=28
#SBATCH --exclusive
#SBATCH --mail-type=BEGIN,END,FAIL
#SBATCH --account=PAS2658
# Applying HaplotypeCaller to do variant calling
/fs/scratch/PASXXXX/Your_WD/Variant_Calling/Sooty_Blotch
# Calling Haplotypes for PD_05d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4.sif gatk --java-options "-Xmx120g" HaplotypeCaller \
    -I PD_05d_sorted_dup_grouped.bam \
    --native-pair-hmm-threads 28 \
    -ERC GVCF \
    -R P_fructicola.fasta \
    -O PD_05d.g.vcf.gz
# For PD_15d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4.sif gatk --java-options "-Xmx120g" HaplotypeCaller \
    -I PD_15d_sorted_dup_grouped.bam \
    --native-pair-hmm-threads 28 \
    -ERC GVCF \
    -R P_fructicola.fasta \
    -O PD_15d.g.vcf.gz
# For For PEG15d
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4.sif gatk --java-options "-Xmx120g" HaplotypeCaller \
    -I PEG15d_sorted_dup_grouped.bam \
    --native-pair-hmm-threads 28 \
    -ERC GVCF \
    -R P_fructicola.fasta \
    -O PEG15d.g.vcf.gz
EOF
sbatch Haplotype_Calling.sh
```

Combine VGCFs
```shell
cd /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Calling
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4.sif gatk --java-options "-Xmx120g" CombineGVCFs \
    -R P_fructicola.fasta \
    -V PD_05d.g.vcf.gz \
    -V PD_15d.g.vcf.gz \
    -V PEG15d.g.vcf.gz \
    -O Sooty_Blotch.g.vcf.gz
```

Apply GenotypeGVCFs
```shell
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4.sif gatk --java-options "-Xmx120g" GenotypeGVCFs \
    -R P_fructicola.fasta \
    -V Sooty_Blotch.g.vcf.gz \
    -O Sooty_Blotch.vcf.gz
```

Using VCFtools for filtering the VCF
```shell
# Keeping only polymorphic SNPs with a minimum depth of 10, no missing data, and quality of 38
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/vcftools.sif vcftools --gzvcf Sooty_Blotch.vcf.gz --remove-indels --maf 0.1 --minDP 10 --max-missing 1 --minQ 38 --recode --out Sooty_Blotch_SNPs_cleaned
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/vcftools.sif vcftools --vcf Sooty_Blotch_SNPs_cleaned.recode.vcf
```

Using VariantsToTable to obtain a tsv file readable in software like Microsoft Excel
```shell
apptainer exec /fs/scratch/PASXXXX/Your_WD/Variant_Calling/Software/GATK4.sif gatk VariantsToTable \
    -R P_fructicola.fasta \
    -V Sooty_Blotch_SNPs_cleaned.recode.vcf \
    -F CHROM -F POS -F FILTER -F TYPE -GF AD -GF DP \
    --show-filtered \
    -O Sooty_Blotch_SNPs_variants.tsv
```
